Index: settings
===================================================================
--- settings
+++ settings
@@ -39,27 +39,13 @@
   },
   "evmVersion": "cancun",
   "viaIR": false,
   "libraries": {
-    "lib/aave-v3-origin/src/contracts/protocol/libraries/logic/BorrowLogic.sol": {
-      "BorrowLogic": "0xFeD9871528E713B5038c4c44BbE7a315f56cAdc6"
-    },
-    "lib/aave-v3-origin/src/contracts/protocol/libraries/logic/ConfiguratorLogic.sol": {
-      "ConfiguratorLogic": "0x6E2aFD57a161d12f34f416c29619BFeAcAC8AA18"
-    },
-    "lib/aave-v3-origin/src/contracts/protocol/libraries/logic/EModeLogic.sol": {
-      "EModeLogic": "0xD1bddC05A3BB5A7907d82A1b4F1E21dBCE69c3d5"
-    },
-    "lib/aave-v3-origin/src/contracts/protocol/libraries/logic/FlashLoanLogic.sol": {
-      "FlashLoanLogic": "0x5e84CEe2afb7B37d2AB14722C39A7c1C26F5B0BB"
-    },
-    "lib/aave-v3-origin/src/contracts/protocol/libraries/logic/LiquidationLogic.sol": {
-      "LiquidationLogic": "0x36Ae486289bB807C3C79A1427b9c3D934294ef43"
-    },
-    "lib/aave-v3-origin/src/contracts/protocol/libraries/logic/PoolLogic.sol": {
-      "PoolLogic": "0xE51B69e5722Bf547866A4d7Bc190c6e81b626806"
-    },
-    "lib/aave-v3-origin/src/contracts/protocol/libraries/logic/SupplyLogic.sol": {
-      "SupplyLogic": "0x034Fd14b9Ae6bB066a1F9f85A55e990b0b25c168"
+    "lib/aave-v3-origin/src/contracts/instances/PoolInstance.sol": {
+      "BorrowLogic": "0x1fb8f7c906cddd28b42ae6eb76abc04a91188635",
+      "FlashLoanLogic": "0xe66adcf153d0bd67a74e5d23e23f90b4011024ec",
+      "LiquidationLogic": "0x8dc095f287dbebd9e15f75bb3dfbf18389bb8104",
+      "PoolLogic": "0x25ed699a052b820d772765e26ed4dd9f0a26f3e5",
+      "SupplyLogic": "0x33654b16a4de97bce05d7dd06803bf1066f3123c"
     }
   }
 }
\ No newline at end of file
