Index: lib/aave-v3-origin/src/contracts/protocol/libraries/configuration/UserConfiguration.sol
===================================================================
--- lib/aave-v3-origin/src/contracts/protocol/libraries/configuration/UserConfiguration.sol
+++ lib/aave-v3-origin/src/contracts/protocol/libraries/configuration/UserConfiguration.sol
@@ -181,11 +181,11 @@
         return self.data & BORROWING_MASK != 0;
     }
 
     /**
-     * @notice Checks if a user has not been using any reserve for borrowing or supply
+     * @notice Checks if a user has not been using any reserve for borrowing, or as collateral
      * @param self The configuration object
-     * @return True if the user has not been borrowing or supplying any reserve, false otherwise
+     * @return True if the user has not been borrowing, or using as collateral any reserve, false otherwise
      */
     function isEmpty(
         DataTypes.UserConfigurationMap memory self
     ) internal pure returns (bool) {
