Index: lib/aave-v3-origin/src/contracts/instances/VariableDebtTokenMainnetInstanceGHO.sol
===================================================================
--- lib/aave-v3-origin/src/contracts/instances/VariableDebtTokenMainnetInstanceGHO.sol
+++ lib/aave-v3-origin/src/contracts/instances/VariableDebtTokenMainnetInstanceGHO.sol
@@ -9,9 +9,9 @@
  * @author BGD Labs
  * @notice Instance of the variable debt token for the Aave protocol for GHO token in Mainnet
  */
 contract VariableDebtTokenMainnetInstanceGHO is VariableDebtToken {
-    uint256 public constant DEBT_TOKEN_REVISION = 5;
+    uint256 public constant DEBT_TOKEN_REVISION = 6;
 
     constructor(
         IPool pool,
         address rewardsController
