Index: settings
===================================================================
--- settings
+++ settings
@@ -29,48 +29,29 @@
     "lib/aave-v3-origin/src/contracts/instances/ATokenInstance.sol": {
       "ATokenInstance": [
         "*"
       ]
-    },
-    "*": {
-      "*": [
-        "abi",
-        "evm.bytecode",
-        "evm.deployedBytecode",
-        "evm.methodIdentifiers",
-        "metadata",
-        "devdoc",
-        "userdoc",
-        "storageLayout",
-        "evm.gasEstimates"
-      ],
-      "": [
-        "ast"
-      ]
     }
   },
   "evmVersion": "shanghai",
   "viaIR": false,
   "libraries": {
     "lib/aave-v3-origin/src/contracts/protocol/libraries/logic/BorrowLogic.sol": {
-      "BorrowLogic": "0xFeD9871528E713B5038c4c44BbE7a315f56cAdc6"
+      "BorrowLogic": "0x1Fb8f7C906cddD28B42ae6Eb76abC04A91188635"
     },
     "lib/aave-v3-origin/src/contracts/protocol/libraries/logic/ConfiguratorLogic.sol": {
       "ConfiguratorLogic": "0x6E2aFD57a161d12f34f416c29619BFeAcAC8AA18"
     },
-    "lib/aave-v3-origin/src/contracts/protocol/libraries/logic/EModeLogic.sol": {
-      "EModeLogic": "0xD1bddC05A3BB5A7907d82A1b4F1E21dBCE69c3d5"
-    },
     "lib/aave-v3-origin/src/contracts/protocol/libraries/logic/FlashLoanLogic.sol": {
-      "FlashLoanLogic": "0x5e84CEe2afb7B37d2AB14722C39A7c1C26F5B0BB"
+      "FlashLoanLogic": "0xE66AdcF153d0bD67a74e5d23e23F90b4011024EC"
     },
     "lib/aave-v3-origin/src/contracts/protocol/libraries/logic/LiquidationLogic.sol": {
-      "LiquidationLogic": "0x36Ae486289bB807C3C79A1427b9c3D934294ef43"
+      "LiquidationLogic": "0x8dc095F287dBEBD9e15f75BB3DfBF18389BB8104"
     },
     "lib/aave-v3-origin/src/contracts/protocol/libraries/logic/PoolLogic.sol": {
-      "PoolLogic": "0xE51B69e5722Bf547866A4d7Bc190c6e81b626806"
+      "PoolLogic": "0x25Ed699a052B820d772765E26Ed4dD9f0A26F3E5"
     },
     "lib/aave-v3-origin/src/contracts/protocol/libraries/logic/SupplyLogic.sol": {
-      "SupplyLogic": "0x034Fd14b9Ae6bB066a1F9f85A55e990b0b25c168"
+      "SupplyLogic": "0x33654b16A4De97bCE05d7dD06803Bf1066f3123C"
     }
   }
 }
\ No newline at end of file
Index: lib/aave-v3-origin/src/contracts/instances/ATokenInstance.sol
===================================================================
--- lib/aave-v3-origin/src/contracts/instances/ATokenInstance.sol
+++ lib/aave-v3-origin/src/contracts/instances/ATokenInstance.sol
@@ -15,9 +15,9 @@
  * @author BGD Labs
  * @notice Instance of the interest bearing token for the Aave protocol
  */
 contract ATokenInstance is AToken {
-    uint256 public constant ATOKEN_REVISION = 4;
+    uint256 public constant ATOKEN_REVISION = 5;
 
     constructor(
         IPool pool,
         address rewardsController,
Index: lib/aave-v3-origin/src/contracts/interfaces/IPool.sol
===================================================================
--- lib/aave-v3-origin/src/contracts/interfaces/IPool.sol
+++ lib/aave-v3-origin/src/contracts/interfaces/IPool.sol
@@ -588,17 +588,15 @@
      * @param from The user from which the aTokens are transferred
      * @param to The user receiving the aTokens
      * @param scaledAmount The scaled amount being transferred/withdrawn
      * @param scaledBalanceFromBefore The aToken scaled balance of the `from` user before the transfer
-     * @param scaledBalanceToBefore The aToken scaled balance of the `to` user before the transfer
      */
     function finalizeTransfer(
         address asset,
         address from,
         address to,
         uint256 scaledAmount,
-        uint256 scaledBalanceFromBefore,
-        uint256 scaledBalanceToBefore
+        uint256 scaledBalanceFromBefore
     ) external;
 
     /**
      * @notice Returns the list of the underlying assets of all the initialized reserves
@@ -676,8 +674,18 @@
         uint128 borrowableBitmap
     ) external;
 
     /**
+     * @notice Replaces the current eMode ltvzeroBitmap.
+     * @param id The id of the category
+     * @param ltvzeroBitmap The ltvzeroBitmap of the category
+     */
+    function configureEModeCategoryLtvzeroBitmap(
+        uint8 id,
+        uint128 ltvzeroBitmap
+    ) external;
+
+    /**
      * @notice Returns the data of an eMode category
      * @dev DEPRECATED use independent getters instead
      * @param id The id of the category
      * @return The configuration data of the category
@@ -687,8 +695,10 @@
     ) external view returns (DataTypes.EModeCategoryLegacy memory);
 
     /**
      * @notice Returns the label of an eMode category
+     * @dev This function is deprecated and will be removed in a future version.
+     * @custom:deprecated
      * @param id The id of the category
      * @return The label of the category
      */
     function getEModeCategoryLabel(
@@ -722,8 +732,17 @@
         uint8 id
     ) external view returns (uint128);
 
     /**
+     * @notice Returns the ltvzero of an eMode category
+     * @param id The id of the category
+     * @return The ltvzeroBitmap of the category
+     */
+    function getEModeCategoryLtvzeroBitmap(
+        uint8 id
+    ) external view returns (uint128);
+
+    /**
      * @notice Allows a user to use the protocol in eMode
      * @param categoryId The id of the category
      */
     function setUserEMode(uint8 categoryId) external;
@@ -910,13 +929,8 @@
      */
     function getBorrowLogic() external view returns (address);
 
     /**
-     * @notice Gets the address of the external EModeLogic
-     */
-    function getEModeLogic() external view returns (address);
-
-    /**
      * @notice Gets the address of the external LiquidationLogic
      */
     function getLiquidationLogic() external view returns (address);
 
Index: lib/aave-v3-origin/src/contracts/protocol/libraries/helpers/Errors.sol
===================================================================
--- lib/aave-v3-origin/src/contracts/protocol/libraries/helpers/Errors.sol
+++ lib/aave-v3-origin/src/contracts/protocol/libraries/helpers/Errors.sol
@@ -87,12 +87,16 @@
     error CallerNotRiskOrPoolOrEmergencyAdmin(); // 'The caller of the function is not a risk, pool or emergency admin'
     error LiquidationGraceSentinelCheckFailed(); // 'Liquidation grace sentinel validation failed'
     error InvalidGracePeriod(); // Grace period above a valid range
     error InvalidFreezeState(); // Reserve is already in the passed freeze state
+    error InvalidLtvzeroState(); // Reserve is already in the passed ltvzero state
     error NotBorrowableInEMode(); // Asset not borrowable in eMode
     error CallerNotUmbrella(); // The caller of the function is not the umbrella contract
     error ReserveNotInDeficit(); // The reserve is not in deficit
     error MustNotLeaveDust(); // Below a certain threshold liquidators need to take the full position
     error UserCannotHaveDebt(); // Thrown when a user tries to interact with a method that requires a position without debt
     error SelfLiquidation(); // Thrown when a user tries to liquidate themselves
     error CallerNotPositionManager(); // Thrown when the caller has not been enabled as a position manager of the on-behalf-of user
+    error InvalidCollateralInEmode(address reserve, uint256 categoryId); /// Thrown when trying to enter an eMode with an invalid collateral asset
+    error InvalidDebtInEmode(address reserve, uint256 categoryId); /// Thrown when trying to enter an eMode with an invalid debt asset
+    error MustBeEmodeCollateral(address reserve, uint256 categoryId); /// Thrown when trying to configure an asset as eMode-ltvzero that is not an eMode collateral
 }
Index: lib/aave-v3-origin/src/contracts/protocol/tokenization/AToken.sol
===================================================================
--- lib/aave-v3-origin/src/contracts/protocol/tokenization/AToken.sol
+++ lib/aave-v3-origin/src/contracts/protocol/tokenization/AToken.sol
@@ -241,9 +241,14 @@
             owner == ECDSA.recover(digest, v, r, s),
             Errors.InvalidSignature()
         );
         _nonces[owner] = currentValidNonce + 1;
-        _approve(owner, spender, value);
+        _approve({
+            owner: owner,
+            spender: spender,
+            amount: value,
+            emitEvent: true
+        });
     }
 
     /// @inheritdoc IERC20
     function transferFrom(
@@ -312,9 +317,8 @@
 
         uint256 index = POOL.getReserveNormalizedIncome(underlyingAsset);
 
         uint256 scaledBalanceFromBefore = super.balanceOf(from);
-        uint256 scaledBalanceToBefore = super.balanceOf(to);
         uint256 scaledAmount = uint256(amount).getATokenTransferScaledAmount(
             index
         );
 
@@ -330,10 +334,9 @@
             asset: underlyingAsset,
             from: from,
             to: to,
             scaledAmount: scaledAmount,
-            scaledBalanceFromBefore: scaledBalanceFromBefore,
-            scaledBalanceToBefore: scaledBalanceToBefore
+            scaledBalanceFromBefore: scaledBalanceFromBefore
         });
     }
 
     /**
Index: lib/aave-v3-origin/src/contracts/protocol/libraries/types/DataTypes.sol
===================================================================
--- lib/aave-v3-origin/src/contracts/protocol/libraries/types/DataTypes.sol
+++ lib/aave-v3-origin/src/contracts/protocol/libraries/types/DataTypes.sol
@@ -144,8 +144,9 @@
         uint16 liquidationBonus;
         uint128 collateralBitmap;
         string label;
         uint128 borrowableBitmap;
+        uint128 ltvzeroBitmap; // if true, the asset will be treated as ltv0 and ltv0 rules apply
     }
 
     enum InterestRateMode {
         NONE,
@@ -188,8 +189,9 @@
         address interestRateStrategyAddress;
         uint256 amount;
         address onBehalfOf;
         uint16 referralCode;
+        uint8 supplierEModeCategory;
     }
 
     struct ExecuteBorrowParams {
         address asset;
@@ -239,9 +241,8 @@
         address from;
         address to;
         uint256 scaledAmount;
         uint256 scaledBalanceFromBefore;
-        uint256 scaledBalanceToBefore;
         address oracle;
         uint8 fromEModeCategory;
     }
 
Index: lib/aave-v3-origin/src/contracts/protocol/tokenization/base/IncentivizedERC20.sol
===================================================================
--- lib/aave-v3-origin/src/contracts/protocol/tokenization/base/IncentivizedERC20.sol
+++ lib/aave-v3-origin/src/contracts/protocol/tokenization/base/IncentivizedERC20.sol
@@ -171,9 +171,14 @@
     function approve(
         address spender,
         uint256 amount
     ) external virtual override returns (bool) {
-        _approve(_msgSender(), spender, amount);
+        _approve({
+            owner: _msgSender(),
+            spender: spender,
+            amount: amount,
+            emitEvent: true
+        });
         return true;
     }
 
     /// @inheritdoc IERC20
@@ -182,58 +187,82 @@
         address recipient,
         uint256 amount
     ) external virtual override returns (bool) {
         uint120 castAmount = amount.toUint120();
-        _approve(
-            sender,
-            _msgSender(),
-            _allowances[sender][_msgSender()] - castAmount
-        );
+        _spendAllowance({
+            owner: sender,
+            spender: _msgSender(),
+            amount: castAmount,
+            correctedAmount: castAmount
+        });
         _transfer(sender, recipient, castAmount);
         return true;
     }
 
     /**
+     * @notice Sets the allowance of the caller to spend `owner`'s tokens to 0.
+     * @param owner The address whose tokens are being renounced.
+     */
+    function renounceAllowance(address owner) external virtual {
+        _approve({
+            owner: owner,
+            spender: _msgSender(),
+            amount: 0,
+            emitEvent: true
+        });
+    }
+
+    /**
      * @notice Increases the allowance of spender to spend _msgSender() tokens
+     * @dev This function is deprecated and will be removed in a future version.
+     * @custom:deprecated
      * @param spender The user allowed to spend on behalf of _msgSender()
      * @param addedValue The amount being added to the allowance
      * @return `true`
      */
     function increaseAllowance(
         address spender,
         uint256 addedValue
     ) external virtual returns (bool) {
-        _approve(
-            _msgSender(),
-            spender,
-            _allowances[_msgSender()][spender] + addedValue
-        );
+        _approve({
+            owner: _msgSender(),
+            spender: spender,
+            amount: _allowances[_msgSender()][spender] + addedValue,
+            emitEvent: true
+        });
         return true;
     }
 
     /**
      * @notice Decreases the allowance of spender to spend _msgSender() tokens
+     * @dev This function is deprecated and will be removed in a future version.
+     * @custom:deprecated
      * @param spender The user allowed to spend on behalf of _msgSender()
      * @param subtractedValue The amount being subtracted to the allowance
      * @return `true`
      */
     function decreaseAllowance(
         address spender,
         uint256 subtractedValue
     ) external virtual returns (bool) {
-        _approve(
-            _msgSender(),
-            spender,
-            _allowances[_msgSender()][spender] - subtractedValue
-        );
+        uint256 currentAllowance = _allowances[_msgSender()][spender];
+
+        _approve({
+            owner: _msgSender(),
+            spender: spender,
+            amount: currentAllowance - subtractedValue,
+            emitEvent: true
+        });
         return true;
     }
 
     /**
      * @dev Updates `owner`'s allowance for `spender` based on spent `value`.
      *
      * Revert if not enough allowance is available.
      *
+     * Doesn't emit the Approval event.
+     *
      * @param owner The owner of the tokens
      * @param spender The user allowed to spend on behalf of owner
      * @param amount The minimum amount being consumed from the allowance
      * @param correctedAmount The maximum amount being consumed from the allowance
@@ -252,12 +281,22 @@
                 amount
             );
         }
 
+        if (currentAllowance == type(uint256).max) {
+            return;
+        }
+
         uint256 consumption = currentAllowance >= correctedAmount
             ? correctedAmount
             : currentAllowance;
-        _approve(owner, spender, currentAllowance - consumption);
+
+        _approve({
+            owner: owner,
+            spender: spender,
+            amount: currentAllowance - consumption,
+            emitEvent: false
+        });
     }
 
     /**
      * @notice Transfers tokens between two users and apply incentives if defined.
@@ -296,16 +335,21 @@
      * @notice Approve `spender` to use `amount` of `owner`s balance
      * @param owner The address owning the tokens
      * @param spender The address approved for spending
      * @param amount The amount of tokens to approve spending of
+     * @param emitEvent Whether to emit the Approval event
      */
     function _approve(
         address owner,
         address spender,
-        uint256 amount
+        uint256 amount,
+        bool emitEvent
     ) internal virtual {
         _allowances[owner][spender] = amount;
-        emit Approval(owner, spender, amount);
+
+        if (emitEvent) {
+            emit Approval(owner, spender, amount);
+        }
     }
 
     /**
      * @notice Update the name of the token
Index: lib/aave-v3-origin/src/contracts/protocol/libraries/math/WadRayMath.sol
===================================================================
--- lib/aave-v3-origin/src/contracts/protocol/libraries/math/WadRayMath.sol
+++ lib/aave-v3-origin/src/contracts/protocol/libraries/math/WadRayMath.sol
@@ -77,17 +77,8 @@
             c := div(add(mul(a, b), HALF_RAY), RAY)
         }
     }
 
-    function rayMul(
-        uint256 a,
-        uint256 b,
-        Rounding rounding
-    ) internal pure returns (uint256 c) {
-        if (rounding == Rounding.Floor) return rayMulFloor(a, b);
-        return rayMulCeil(a, b);
-    }
-
     function rayMulFloor(
         uint256 a,
         uint256 b
     ) internal pure returns (uint256 c) {
@@ -135,17 +126,8 @@
             c := div(add(mul(a, RAY), div(b, 2)), b)
         }
     }
 
-    function rayDiv(
-        uint256 a,
-        uint256 b,
-        Rounding rounding
-    ) internal pure returns (uint256 c) {
-        if (rounding == Rounding.Floor) return rayDivFloor(a, b);
-        return rayDivCeil(a, b);
-    }
-
     function rayDivCeil(
         uint256 a,
         uint256 b
     ) internal pure returns (uint256 c) {
