===================================================================
--- src/contracts/interfaces/IPoolConfigurator.sol
+++ lib/aave-v3-origin/src/contracts/interfaces/IPoolConfigurator.sol
@@ -176,8 +176,20 @@
         bool borrowable
     );
 
     /**
+     * @dev Emitted when the ltvzero configuration of an asset in an eMode changed.
+     * @param asset The address of the underlying asset of the reserve
+     * @param categoryId The eMode category
+     * @param ltvzero True if the asset is enabled as ltvzero in the eMode, false otherwise.
+     */
+    event AssetLtvzeroInEModeChanged(
+        address indexed asset,
+        uint8 categoryId,
+        bool ltvzero
+    );
+
+    /**
      * @dev Emitted when a new eMode category is added or an existing category is altered.
      * @param categoryId The new eMode category id
      * @param ltv The ltv for the asset category in eMode
      * @param liquidationThreshold The liquidationThreshold for the asset category in eMode
@@ -367,14 +379,23 @@
 
     /**
      * @notice Freeze or unfreeze a reserve. A frozen reserve doesn't allow any new supply, borrow
      * or rate swap but allows repayments, liquidations, rate rebalances and withdrawals.
+     * Also sets ltvzero on affected eModes and reserve ltv to 0.
+     * Stores the current reserve ltv in pendingLtv[asset].
      * @param asset The address of the underlying asset of the reserve
      * @param freeze True if the reserve needs to be frozen, false otherwise
      */
     function setReserveFreeze(address asset, bool freeze) external;
 
     /**
+     * @notice Restores the pending ltv for a reserve in non-emode (emode 0), if it is non-zero.
+     * @param asset The address of the underlying asset of the reserve
+     * @param ltvzero True if the reserve should be flagged as ltvzero, false otherwise
+     */
+    function setReserveLtvzero(address asset, bool ltvzero) external;
+
+    /**
      * @notice Sets the borrowable in isolation flag for the reserve.
      * @dev When this flag is set to true, the asset will be borrowable against isolated collaterals and the
      * borrowed amount will be accumulated in the isolated collateral's total debt exposure
      * @dev Only assets of the same family (e.g. USD stablecoins) should be borrowable in isolation mode to keep
@@ -475,9 +496,8 @@
     function setLiquidationProtocolFee(address asset, uint256 newFee) external;
 
     /**
      * @notice Enables/disables an asset to be borrowable in a selected eMode.
-     * - eMode.borrowable always has less priority then reserve.borrowable
      * @param asset The address of the underlying asset of the reserve
      * @param categoryId The eMode categoryId
      * @param borrowable True if the asset should be borrowable in the given eMode category, false otherwise.
      */
@@ -499,8 +519,20 @@
         bool collateral
     ) external;
 
     /**
+     * @notice Enables/disables an asset to be collateral in a selected eMode.
+     * @param asset The address of the underlying asset of the reserve
+     * @param categoryId The eMode categoryId
+     * @param ltvzero True if the asset should be ltvzero in the given eMode category, false otherwise.
+     */
+    function setAssetLtvzeroInEMode(
+        address asset,
+        uint8 categoryId,
+        bool ltvzero
+    ) external;
+
+    /**
      * @notice Adds a new efficiency mode (eMode) category or alters a existing one.
      * @param categoryId The id of the category to be configured
      * @param ltv The ltv associated with the category
      * @param liquidationThreshold The liquidation threshold associated with the category
