Index: lib/aave-v3-origin/src/contracts/instances/ATokenWithDelegationInstance.sol
===================================================================
--- lib/aave-v3-origin/src/contracts/instances/ATokenWithDelegationInstance.sol
+++ lib/aave-v3-origin/src/contracts/instances/ATokenWithDelegationInstance.sol
@@ -15,9 +15,9 @@
  * @author BGD Labs
  * @notice Instance of the interest bearing token for the Aave protocol with additional delegation functionality used on aAAVE(mainnet)
  */
 contract ATokenWithDelegationInstance is ATokenWithDelegation {
-    uint256 public constant ATOKEN_REVISION = 4;
+    uint256 public constant ATOKEN_REVISION = 5;
 
     constructor(
         IPool pool,
         address rewardsController,
